druid/src/widget/images.rs

```diff
@@ -0,0 +1,259 @@
+    let origin = Point::new(origin_x, origin_y);
+
+    (scale, origin)
+}
+
+/// A widget that renders a Images
+pub struct Image<T> {
+    image_data: ImageData,
+    phantom: PhantomData<T>,
+
+    vert: FloatVert,
+    hoz: FloatHos,
+    fill: FillStrat,
+}
+
+impl<T: Data> Image<T> {
+    /// Create an Image-drawing widget from ImageData.
+    ///
+    /// The Image will scale to fit its box constraints.
+    pub fn new(image_data: ImageData, fillstr: FillStrat) -> impl Widget<T> {
+        Image {
+            image_data,
+            phantom: Default::default(),
+            vert: FloatVert::Center,
+            hoz: FloatHos::Center,
+            fill: fillstr,
+        }
+    }
+
+    /// Measure the Image's size
+    #[allow(clippy::needless_return)]
+    fn get_size(&self) -> Size {
+        // Fix me
+        Size::new(
+            self.image_data.x_pixels as f64,
+            self.image_data.y_pixels as f64,
+        )
+    }
+
+    pub fn set_fill(&mut self, newfil: FillStrat) {
+        self.fill = newfil;
+    }
+}
+
+impl<T: Data> Widget<T> for Image<T> {
+    fn event(&mut self, _ctx: &mut EventCtx, _event: &Event, _data: &mut T, _env: &Env) {}
+
+    fn lifecycle(&mut self, _ctx: &mut LifeCycleCtx, _event: &LifeCycle, _data: &T, _env: &Env) {}
+
+    fn update(&mut self, _ctx: &mut UpdateCtx, _old_data: &T, _data: &T, _env: &Env) {}
+
+    fn layout(
+        &mut self,
+        _layout_ctx: &mut LayoutCtx,
+        bc: &BoxConstraints,
+        _data: &T,
+        _env: &Env,
+    ) -> Size {
+        bc.debug_check("Image");
+
+        if bc.is_width_bounded() {
+            bc.max()
+        } else {
+            bc.constrain(self.get_size())
+        }
+    }
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, _data: &T, _env: &Env) {
+        let bob = get_scale_offset(
+            paint_ctx.size(),
+            self.get_size(),
+            &self.fill,
+            &self.hoz,
+            &self.vert,
+        );
+
+        // The ImageData's to_piet function dose not clip to the image's size
+        // CairoRenderContext is very like druids but with some extra goodies like clip
+        if self.fill == FillStrat::Contain {
+        } else {
+            let clip_rect = Rect::ZERO.with_size(paint_ctx.size());
+            paint_ctx.clip(clip_rect);
+        }
+        self.image_data.to_piet(bob.0.x, bob.1, paint_ctx);
+    }
+}
+
+/// Stored Image data.
+/// Implements `FromStr` and can be converted to piet draw instructions.
+#[derive(Clone)]
+pub struct ImageData {
+    pixels: Vec<u8>,
+    x_pixels: i32,
+    y_pixels: i32,
+}
+
+impl ImageData {
+    /// Create an empty Image
+    pub fn empty() -> Self {
+        ImageData {
+            pixels: [].to_vec(),
+            x_pixels: 0,
+            y_pixels: 0,
+        }
+    }
+
+    pub fn from_data(raw_image: Vec<u8>) -> Self {
+        let dec = image::load_from_memory(&raw_image[..]).unwrap().to_rgb();
```

Comment:
this probably shouldn't unwrap, we don't know where this data is coming from; I think it is reasonable that we return a `Result` here.

Also I think we can make this take a `&[u8]`, which might save a clone.